### PR TITLE
Enhance object state handling for AbstractPath

### PR DESCRIPTION
**Pull request by RaphaelLilt**

### CODE REVIEW

File: sisyphus/job_path.py [hunk 1]
```diff
@@ -239,12 +239,19 @@ def __hash__(self):
         # TODO Check how uninitialized object should behave here
         return hash((self.__dict__.get("creator"), self.__dict__.get("path")))
 
+    def _get_state_with_creator(self):
+        d = self.__dict__.copy()
+        del d["users"]
+        return d
```
Comment (Numeri) [inline sisyphus/job_path.py:245]: Can you add a comment describing why we need this function (for creating accurate hashes) and why we still need to delete `users`?

Comment (RaphaelLilt) [reply sisyphus/job_path.py:245]: The users were already deleted in the old code. Not sure why.

Will add a comment about the hashes.

**Review by Numeri — APPROVED** [review]
LGTM

**Comment by gabrielbretschner** [comment]
did you test this with a clean setup (ie no jobs have run yet)? From what I can tell, the function `__get_state__` is also used when traversing the graph to calculate the required inputs set.

If you look at `Job._sis_init` it uses `tools.extract_paths` to detect all inputs 
```
        self._sis_inputs.update({p for p in tools.extract_paths([self.__dict__, args, kwargs]) if p.creator != self})
```
this method iterates over all the args recursively until it either ignored all items (ie just basic variables) or it calls `_sis_path` on an object if it has it. For this to work it uses `get_object_state` which sometimes uses `__get_state__`.

This might not be an issue at all. I just wanted to document this case as I think its a bit more obscure. 


**Update**: this might be an handled issue in your branch. I was looking at the orginal code but it seems in `get_object_state` you do use the version with creator

**Comment by RaphaelLilt** [comment]
> did you test this with a clean setup (ie no jobs have run yet)? From what I can tell, the function `__get_state__` is also used when traversing the graph to calculate the required inputs set.
> 
> If you look at `Job._sis_init` it uses `tools.extract_paths` to detect all inputs
> 
> ```
>         self._sis_inputs.update({p for p in tools.extract_paths([self.__dict__, args, kwargs]) if p.creator != self})
> ```
> 
> this method iterates over all the args recursively until it either ignored all items (ie just basic variables) or it calls `_sis_path` on an object if it has it. For this to work it uses `get_object_state` which sometimes uses `__get_state__`.
> 
> This might not be an issue at all. I just wanted to document this case as I think its a bit more obscure.
> 
> **Update**: this might be an handled issue in your branch. I was looking at the orginal code but it seems in `get_object_state` you do use the version with creator

Yes, that's handled here:  https://github.com/RaphaelLilt/sisyphus/blob/a18662881f4878c949f9de8e83af5f0af5d954e3/sisyphus/hash.py#L72

I'm not sure if we can also optimize this by not including the creator of an AbstractPath but that needs more investigation.

**Comment by gabrielbretschner** [comment]
The way you implemented it seems safe. Could you open a PR to the upstream repo to get more input on this?

**Pull request closed without merging**